### DISCUSSION TITLE
Using Ice in games?

### DISCUSSION THREAD

**Discussion by ethindp** [zeroc-ice/ice#1429] [Q&A] [answered]
Hi all,
I was thinking about what using Ice in an online game would look like, and what immediately came to mind was a pretty significant feature that looks like it's missing, though maybe it's just hidden somewhere: the ability to make the communicator 3-way/4-way, allowing the server to call functions defined in the client.

Games use this trick for all kinds of things: broadcasting things to all connected clients (which is also something that I can't seem to find), sending notifications to individual clients, etc. And it's not just Games that do this, either. It's messaging/chat apps, and so on and so forth. I suppose that this is probably a pretty unorthodox way of using RPC; I get the feeling that RPC is usually just used for client-server communication, but the overall idea that, hey, we can define a fully type-safe communication architecture (with all the features that Ice has) is extremely tantalizing. Would this be doable in Ice? The manual says nothing about this, so I get the idea that this isn't something done under normal circumstances, but I thought I'd ask anyway.

**Comment by ethindp** [comment]
Update: I think I figured it out, but are there any caveats to game-specific uses that I should be aware of (other than it most likely only supporting TCP/WS/WSS/BT/IAP)?

**Comment by pepone** [comment, marked as answer]
Ice supports bidirectional communications which allow the server to call back the client over the connection initiated by the client.

see https://doc.zeroc.com/ice/3.7/client-server-features/connection-management/bidirectional-connections

If you need to pass the same message to multiple peers one option is to use a publish-subscriber framework like [DataStorm]( https://doc.zeroc.com/datastorm/latest/introduction) or [IceStorm](https://doc.zeroc.com/ice/3.7/ice-services/icestorm)

**Comment by ethindp** [comment]
Thank you. I knew I was missing something.
